File: cookbook/ralph-loop.py
"""
ralph-loop.py — CopilotForge Cookbook Recipe

WHAT THIS DOES:
    Implements the autonomous dev loop pattern (Ralph Loop): read a plan
    from disk, pick the next pending task, implement it, validate the result,
    mark it done (with git commit) or failed, and repeat. State lives on
    disk — not in memory — so each iteration starts with fresh context.

WHEN TO USE THIS:
    When you want an agent to work through an implementation plan
    autonomously — picking tasks, writing code, validating, and committing
    without human intervention between steps.

HOW TO RUN:
    1. Create an IMPLEMENTATION_PLAN.md in your project root (see format below)
    2. python cookbook/ralph-loop.py

PREREQUISITES:
    - Python 3.10+
    - git initialized in the project
    - An IMPLEMENTATION_PLAN.md file (format shown in code)

EXPECTED OUTPUT:
    [Ralph] Loaded 3 tasks from IMPLEMENTATION_PLAN.md
    [Ralph] === Iteration 1/10 ===
    [Ralph] Picked task: add-utils — "Create utility helpers"
    [Ralph] Implementing: add-utils...
    [Ralph] Validating: add-utils...
    [Ralph] ✅ Task add-utils passed — committing.
    [Ralph] === Iteration 2/10 ===
    ...
    [Ralph] 🏁 All tasks complete. 3 done, 0 failed.

PLATFORM NOTES:
    - Windows: Use backslashes in paths or os.path.join() (both shown in code)
    - macOS/Linux: Forward slashes work natively
    - Environment variables: Use $env:VAR (PowerShell) or export VAR (bash)
"""

from __future__ import annotations

import os
import re
import subprocess
from dataclasses import dataclass
from pathlib import Path


# --- Types ---


@dataclass
class Task:
    """A single task from the implementation plan."""

    id: str
    title: str
    status: str  # "pending", "done", or "failed"


# --- Plan Parser ---


def parse_plan(file_path: Path) -> list[Task]:
    """
    Parse IMPLEMENTATION_PLAN.md into a list of tasks.
    Expected format — one task per line:
        - [ ] task-id — Task title
        - [x] task-id — Task title        (done)
        - [!] task-id — Task title        (failed)
    """
    content = file_path.read_text(encoding="utf-8")
    tasks: list[Task] = []

    for line in content.splitlines():
        match = re.match(r"^- \[(.)\] (\S+)\s*—\s*(.+)$", line)
        if not match:
            continue

        marker, task_id, title = match.group(1), match.group(2), match.group(3).strip()
        status = "pending"
        if marker == "x":
            status = "done"
        elif marker == "!":
            status = "failed"

        tasks.append(Task(id=task_id, title=title, status=status))

    return tasks


def write_plan(file_path: Path, tasks: list[Task]) -> None:
    """Write the task list back to IMPLEMENTATION_PLAN.md."""
    lines = ["# Implementation Plan", ""]
    for task in tasks:
        marker = {"done": "x", "failed": "!", "pending": " "}.get(task.status, " ")
        lines.append(f"- [{marker}] {task.id} — {task.title}")
    file_path.write_text("\n".join(lines) + "\n", encoding="utf-8")


# --- Implementation (simulated) ---


def implement_task(task: Task) -> None:
    """
    Simulate implementing a task — creates or modifies a file.
    TODO: Replace with actual Copilot SDK call to generate code.
    """
    out_dir = Path("src")
    out_dir.mkdir(parents=True, exist_ok=True)

    file_path = out_dir / f"{task.id}.py"
    content = f'# Auto-generated for task: {task.id}\n# {task.title}\n'
    file_path.write_text(content, encoding="utf-8")
    print(f"[Ralph] Implementing: {task.id}...")


# --- Validation (simulated) ---


def validate_task(task: Task) -> bool:
    """
    Simulate validation — run tests or a build check.
    TODO: Replace with actual validation (e.g., subprocess.run(["pytest"])).
    """
    print(f"[Ralph] Validating: {task.id}...")
    return True


# --- Git Helpers ---


def git_commit(message: str) -> None:
    """Stage all changes and commit."""
    try:
        subprocess.run(["git", "add", "-A"], capture_output=True, check=True)
        subprocess.run(
            ["git", "commit", "-m", message, "--allow-empty"],
            capture_output=True,
            check=True,
        )
    except (subprocess.CalledProcessError, FileNotFoundError):
        print("[Ralph] Git commit skipped (no changes or git not configured).")


# --- Ralph Loop ---


def ralph_loop(plan_path: Path, max_iterations: int = 10) -> None:
    """Main loop: pick task → implement → validate → commit → repeat."""
    if not plan_path.exists():
        print(f"[Ralph] Plan not found: {plan_path}")
        print("[Ralph] Create an IMPLEMENTATION_PLAN.md with tasks like:")
        print("  - [ ] task-id — Task title")
        return

    iteration = 0

    while iteration < max_iterations:
        iteration += 1
        # Fresh read from disk on every iteration (key principle).
        tasks = parse_plan(plan_path)

        if iteration == 1:
            print(f"[Ralph] Loaded {len(tasks)} tasks from {plan_path}")

        pending = next((t for t in tasks if t.status == "pending"), None)
        if pending is None:
            done = sum(1 for t in tasks if t.status == "done")
            failed = sum(1 for t in tasks if t.status == "failed")
            print(f"[Ralph] 🏁 All tasks complete. {done} done, {failed} failed.")
            return

        print(f"[Ralph] === Iteration {iteration}/{max_iterations} ===")
        print(f'[Ralph] Picked task: {pending.id} — "{pending.title}"')

        # Implement
        implement_task(pending)

        # Validate
        passed = validate_task(pending)

        # Update status on disk
        fresh_tasks = parse_plan(plan_path)
        target = next((t for t in fresh_tasks if t.id == pending.id), None)
        if target:
            target.status = "done" if passed else "failed"
            write_plan(plan_path, fresh_tasks)

        if passed:
            print(f"[Ralph] ✅ Task {pending.id} passed — committing.")
            git_commit(f"feat: {pending.id} — {pending.title}")
        else:
            print(f"[Ralph] ❌ Task {pending.id} failed — logged.")

    print(f"[Ralph] ⚠️ Reached max iterations ({max_iterations}). Stopping.")


# --- Entry Point ---

if __name__ == "__main__":
    plan_file = Path("IMPLEMENTATION_PLAN.md")
    ralph_loop(plan_file)

File: cookbook/ralph-loop.ts
#!/usr/bin/env tsx
/**
 * ralph-loop.ts — CopilotForge Development Tool
 *
 * WHAT THIS DOES:
 *   Autonomous development loop runner for CopilotForge. Reads a prompt
 *   file from the project root (PROMPT_plan.md or PROMPT_build.md) and
 *   drives a Copilot session in a loop until all tasks are done or the
 *   iteration cap is reached. Every tool call is logged (⚙ toolName) so
 *   progress is always visible.
 *
 * WHEN TO USE THIS:
 *   When you want to run CopilotForge's development cycle without sitting
 *   at the keyboard. Choose plan mode to populate the backlog, build mode
 *   to work through it. Sessions are fully autonomous — no permission
 *   prompts, no interruptions.
 *
 * HOW TO RUN:
 *   1. npm install @github/copilot-sdk
 *   2. Create PROMPT_plan.md or PROMPT_build.md at the project root
 *   3. npx tsx cookbook/ralph-loop.ts [plan|build] [max_iterations]
 *
 *   Examples:
 *     npx tsx cookbook/ralph-loop.ts build        # build mode, 50 iterations
 *     npx tsx cookbook/ralph-loop.ts plan 10      # plan mode, 10 iterations
 *     npx tsx cookbook/ralph-loop.ts build 5      # build mode, 5 iterations
 *
 * PREREQUISITES:
 *   - Node.js 18+
 *   - TypeScript 5+ (tsx or ts-node for direct execution)
 *   - @github/copilot-sdk (npm install @github/copilot-sdk)
 *   - PROMPT_plan.md or PROMPT_build.md at the project root
 *
 * EXPECTED OUTPUT:
 *   🔄 Ralph Loop — build mode — max 50 iterations
 *   📁 Working directory: /your/project/root
 *
 *   --- Iteration 1/50 [build mode] ---
 *     ⚙ read_file
 *     ⚙ edit_file
 *     ⚙ run_command
 *     ✓ Iteration complete
 *
 *   --- Iteration 2/50 [build mode] ---
 *     ⚙ git_commit
 *     ✓ Iteration complete
 *
 *   ✅ All tasks complete. Ralph Loop exiting.
 *   🔄 Ralph Loop finished — build mode — 50 iterations max
 *
 * PLATFORM NOTES:
 *   - Windows: Use `npx tsx` rather than `npx ts-node` for best results
 *   - macOS/Linux: Both tsx and ts-node work
 *   - Environment variables: Use $env:VAR (PowerShell) or export VAR (bash)
 *
 * NOTE: This file is a CopilotForge development tool — it is NOT scaffolded
 *   into user repos by `copilotforge init`. It lives only in cookbook/.
 */

// TODO: Replace mock types with the real import once @github/copilot-sdk is installed:
//   import { CopilotClient } from "@github/copilot-sdk";
//   npm install @github/copilot-sdk

import { readFileSync, existsSync } from "node:fs";
import { join } from "node:path";

// --- Types ---

/** A tool-use event emitted by the session as it works. */
interface ToolCallEvent {
  name: string;
}

/** The callback fired when the session wants to use a tool. */
type ToolCallHandler = (toolName: string) => void;

/** Passed to onPermissionRequest; contains tool name and arguments. */
interface PermissionRequest {
  toolName: string;
  args?: Record<string, unknown>;
}

/** Return this from onPermissionRequest to grant or deny tool use. */
interface PermissionResponse {
  approved: boolean;
}

/** Options forwarded to session.sendAndWait. */
interface SendOptions {
  onToolCall?: ToolCallHandler;
}

/** A single autonomous Copilot session. */
interface CopilotSession {
  sendAndWait(prompt: string, options?: SendOptions): Promise<string>;
}

/** Configuration for creating a CopilotSession. */
interface SessionConfig {
  workingDirectory: string;
  onPermissionRequest: (req: PermissionRequest) => Promise<PermissionResponse>;
}

/** Top-level SDK client. */
interface CopilotClientInterface {
  createSession(config: SessionConfig): Promise<CopilotSession>;
}

// TODO: Remove this mock class once @github/copilot-sdk is installed.
//       Replace instantiation below with: const client = new CopilotClient();
class MockCopilotClient implements CopilotClientInterface {
  async createSession(config: SessionConfig): Promise<CopilotSession> {
    // Simulate a session that auto-approves and echoes tool call events.
    return {
      async sendAndWait(prompt: string, options?: SendOptions): Promise<string> {
        // Simulate two tool calls per iteration so the log output is meaningful.
        const simulatedTools = ["read_file", "edit_file"];
        for (const tool of simulatedTools) {
          options?.onToolCall?.(tool);
          await new Promise((r) => setTimeout(r, 50));
        }
        // Return a non-terminal response so the loop runs to max iterations.
        return "Working on next task…";
      },
    };
  }
}

// --- Prompt loader ---

/**
 * Reads the mode-specific prompt file from the project root.
 * Throws if the file is missing so the user gets a clear message.
 */
function loadPrompt(mode: "plan" | "build", projectRoot: string): string {
  const fileName = mode === "plan" ? "PROMPT_plan.md" : "PROMPT_build.md";
  const filePath = join(projectRoot, fileName);

  if (!existsSync(filePath)) {
    throw new Error(
      `Prompt file not found: ${filePath}\n` +
        `Create ${fileName} at the project root before running ralph-loop.`
    );
  }

  return readFileSync(filePath, "utf-8").trim();
}

// --- Completion detector ---

/**
 * Returns true when the session response signals that no more tasks remain.
 * Extend this list as the project adds new completion signals.
 */
function isComplete(response: string): boolean {
  const signals = [
    "NO_MORE_TASKS",
    "Board is clear",
    "DONE",
    "All tasks complete",
    "nothing left to do",
  ];
  const lower = response.toLowerCase();
  return signals.some((s) => lower.includes(s.toLowerCase()));
}

// --- Main loop ---

async function runLoop(): Promise<void> {
  const mode = (process.argv[2] ?? "build") as "plan" | "build";
  const maxIterations = parseInt(process.argv[3] ?? "50", 10);
  const projectRoot = process.cwd();

  if (mode !== "plan" && mode !== "build") {
    console.error(`❌ Unknown mode "${mode}". Use "plan" or "build".`);
    process.exit(1);
  }

  if (isNaN(maxIterations) || maxIterations < 1) {
    console.error(`❌ max_iterations must be a positive integer. Got: ${process.argv[3]}`);
    process.exit(1);
  }

  console.log(`🔄 Ralph Loop — ${mode} mode — max ${maxIterations} iterations`);
  console.log(`📁 Working directory: ${projectRoot}\n`);

  let prompt: string;
  try {
    prompt = loadPrompt(mode, projectRoot);
  } catch (err) {
    console.error(`❌ ${err instanceof Error ? err.message : err}`);
    process.exit(1);
  }

  // TODO: Replace MockCopilotClient with the real SDK:
  //   import { CopilotClient } from "@github/copilot-sdk";
  //   const client = new CopilotClient();
  const client: CopilotClientInterface = new MockCopilotClient();

  let tasksCompleted = 0;
  let lastCommitMessage = "(none)";

  for (let i = 1; i <= maxIterations; i++) {
    console.log(`\n--- Iteration ${i}/${maxIterations} [${mode} mode] ---`);

    try {
      const session = await client.createSession({
        workingDirectory: projectRoot,
        // Auto-approve every tool call — no interactive prompts.
        onPermissionRequest: async (_req: PermissionRequest): Promise<PermissionResponse> => ({
          approved: true,
        }),
      });

      const response = await session.sendAndWait(prompt, {
        onToolCall: (toolName: string) => {
          console.log(`  ⚙ ${toolName}`);
          // Track git commits for the exit summary.
          if (toolName === "git_commit" || toolName === "create_commit") {
            tasksCompleted += 1;
            lastCommitMessage = `commit detected at iteration ${i}`;
          }
        },
      });

      console.log(`  ✓ Iteration complete`);

      if (isComplete(response)) {
        console.log("\n✅ All tasks complete. Ralph Loop exiting.");
        break;
      }
    } catch (err) {
      console.error(
        `  ❌ Iteration ${i} failed:`,
        err instanceof Error ? err.message : err
      );
      // Log and continue — never crash the entire loop on a single bad iteration.
    }
  }

  console.log(`\n🔄 Ralph Loop finished — ${mode} mode — ${maxIterations} iterations max`);
  console.log(`   Tasks completed (commits detected): ${tasksCompleted}`);
  console.log(`   Last commit note: ${lastCommitMessage}`);
}

runLoop().catch(console.error);
